contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,683 @@
+
+	json_node_free(root);
+	g_object_unref(gen);
+
+	res = TRUE;
+
+out:
+	g_object_unref(builder);
+	g_free(description);
+	g_free(operation);
+	g_free(last_error);
+	g_free(header_value);
+	g_free(json_str);
+
+	return res;
+}
+
+/*
+ * Generate JSON containing status information about slots and print it to
+ * stdout.
+ */
+static gboolean slot_status(RInstaller *installer, GError **error)
+{
+	JsonBuilder *builder = json_builder_new();
+	JsonGenerator *gen = NULL;
+	JsonNode *root = NULL;
+	GVariantIter iter_slots;
+	GVariantIter iter_props;
+	GVariant *slot_status_array = NULL;
+	GVariant *properties = NULL;
+	GVariant *value = NULL;
+	gchar *key = NULL;
+	gchar *slot_name = NULL;
+	const gchar *property_str = NULL;
+	guint32 property_u32;
+	guint64 property_u64;
+	gboolean res = FALSE;
+
+	res = r_installer_call_get_slot_status_sync(installer, &slot_status_array, NULL, error);
+	if (!res)
+		goto out;
+
+	json_builder_begin_object(builder);
+
+	g_variant_iter_init(&iter_slots, slot_status_array);
+
+	/* iterate slots */
+	while ((g_variant_iter_next(&iter_slots, "(s@a{sv})", &slot_name, &properties))) {
+		json_builder_set_member_name(builder, slot_name);
+		json_builder_begin_object(builder);
+
+		g_variant_iter_init(&iter_props, properties);
+
+		/* iterate attributes */
+		while ((g_variant_iter_next(&iter_props, "{sv}", &key, &value))) {
+			json_builder_set_member_name(builder, key);
+
+			if (g_variant_is_of_type(value, G_VARIANT_TYPE_STRING)) {
+				property_str = g_variant_get_string(value, NULL);
+				json_builder_add_string_value(builder, property_str);
+			}
+			if (g_variant_is_of_type(value, G_VARIANT_TYPE_UINT32)) {
+				property_u32 = g_variant_get_uint32(value);
+				json_builder_add_int_value(builder, property_u32);
+			}
+			if (g_variant_is_of_type(value, G_VARIANT_TYPE_UINT64)) {
+				property_u64 = g_variant_get_uint64(value);
+				json_builder_add_int_value(builder, property_u64);
+			}
+			/*
+			 * ignore all other types as they are unknown at the
+			 * time of writing
+			 */
+		}
+
+		json_builder_end_object(builder);
+	}
+
+	json_builder_end_object(builder);
+
+	gen = json_generator_new();
+	root = json_builder_get_root(builder);
+	json_generator_set_root(gen, root);
+
+	print_headers("200 OK", "application/json");
+	g_print("%s\n", json_generator_to_data(gen, NULL));
+
+	json_node_free(root);
+	g_object_unref(builder);
+	g_variant_unref(value);
+	g_variant_unref(properties);
+	g_variant_unref(slot_status_array);
+
+out:
+	if (gen)
+		g_object_unref(gen);
+	g_free(slot_name);
+	g_free(key);
+
+	return res;
+}
+
+/*
+ * Wrapper around fread() that writes upload progress information to status
+ * file.
+ */
+static gint64 fread_chunked(char *ptr, size_t size, FILE *stream, GError **error)
+{
+	gint64 items_read = 0;
+	gint64 bytes_read = 0;
+	gint percentage = 0;
+	gint item_size = 8;
+
+	while (bytes_read < (gint64) size) {
+		items_read = fread(ptr + bytes_read, item_size, 1024, stream);
```

Comment:
This will not work except for files which have a size that is multiple of item_size (8) bytes!

Either set `item_size` to 1 or use an appropriate glib method

Reply:
Oops, thanks for catching this.